core/optaplanner-constraint-streams/src/main/java/org/optaplanner/constraint/streams/bavet/BavetConstraintSession.java

```diff
@@ -16,83 +16,38 @@
-                constraintMatchEnabled);
-        declaredClassToNodeMap = new HashMap<>(50);
-        BavetNodeBuildPolicy<Solution_> buildPolicy = new BavetNodeBuildPolicy<>(this);
-        constraintToWeightMap.forEach((constraint, constraintWeight) -> constraint.createNodes(buildPolicy,
-                declaredClassToNodeMap, constraintWeight));
-        nodeIndexedNodeMap = buildPolicy.getCreatedNodes();
-        scoringNodeList = nodeIndexedNodeMap.stream()
-                .filter(node -> node instanceof BavetScoringNode)
-                .map(node -> (BavetScoringNode) node)
-                .collect(Collectors.toList());
-        effectiveClassToNodeListMap = new HashMap<>(declaredClassToNodeMap.size());
-        int nodeCount = nodeIndexedNodeMap.size();
-        nodeIndexToDirtyTupleQueueMap = new ArrayList<>(nodeCount);
-        for (int i = 0; i < nodeCount; i++) {
-            nodeIndexToDirtyTupleQueueMap.add(new ArrayDeque<>(1000));
-        }
-        fromTupleListMap = new IdentityHashMap<>(1000);
-    }
+    private final Map<Class<?>, ForEachUniNode<Object>> declaredClassToNodeMap;
+    private final AbstractNode[] nodes; // Indexed by nodeIndex
 
-    private static void refreshTuple(BavetAbstractTuple tuple) {
-        tuple.getNode().refresh(tuple);
-        switch (tuple.getState()) {
-            case CREATING:
-            case UPDATING:
-                tuple.setState(BavetTupleState.OK);
-                return;
-            case DYING:
-            case ABORTING:
-                tuple.setState(BavetTupleState.DEAD);
-                return;
-            case DEAD:
-                throw new IllegalStateException("Impossible state: The tuple (" + tuple + ") in node (" +
-                        tuple.getNode() + ") is already in the dead state (" + tuple.getState() + ").");
-            default:
-                throw new IllegalStateException("Impossible state: Tuple (" + tuple + ") in node (" +
-                        tuple.getNode() + ") is in an unexpected state (" + tuple.getState() + ").");
-        }
+    private final Map<Class<?>, List<ForEachUniNode<Object>>> effectiveClassToNodeListMap;
+
+    public BavetConstraintSession(AbstractScoreInliner<Score_> scoreInliner,
+            Map<Class<?>, ForEachUniNode<Object>> declaredClassToNodeMap,
+            AbstractNode[] nodes) {
+        this.scoreInliner = scoreInliner;
+        this.declaredClassToNodeMap = declaredClassToNodeMap;
+        this.nodes = nodes;
+        effectiveClassToNodeListMap = new HashMap<>(declaredClassToNodeMap.size());
     }
 
-    public List<BavetFromUniNode<Object>> findFromNodeList(Class<?> factClass) {
+    public List<ForEachUniNode<Object>> findNodeList(Class<?> factClass) {
         return effectiveClassToNodeListMap.computeIfAbsent(factClass, key -> {
-            List<BavetFromUniNode<Object>> nodeList = new ArrayList<>();
+            List<ForEachUniNode<Object>> nodeList = new ArrayList<>();
```

Comment:
Maybe there is an opportunity to start using the new bits from JDK 11?

See this method, for example:
https://docs.oracle.com/en/java/javase/11/docs/api/java.base/java/util/List.html#copyOf(java.util.Collection)

Will take your array list and create an immutable copy which is properly sized. More importantly, if the list has 1 item, it will automatically create a smart list which is not actually even a list, just a pointer to one element. Memory savings galore.

Reply:
Interesting to know - we can probably use that on other places (too).

In this case, we need a filter (predicate) for the isAssignableFrom, so we can't use copyOf I think. But we could do a stream - of course. Not sure if that's better or not than our current impl.